zk_dtypes/include/field/frobenius.h

```diff
@@ -0,0 +1,72 @@
+/* Copyright 2025 The zk_dtypes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+#ifndef ZK_DTYPES_INCLUDE_FIELD_FROBENIUS_H_
+#define ZK_DTYPES_INCLUDE_FIELD_FROBENIUS_H_
+
+#include <array>
+#include <type_traits>
+
+#include "zk_dtypes/include/field/extension_field_operation_traits_forward.h"
+
+namespace zk_dtypes {
+
+// Frobenius endomorphism: φ(x) = x^p
+//
+// For extension field Fp[u] / (u^n - ξ):
+//   φ(x_k · u^k) = φ(x_k) · u^{kp mod n} · ξ^{kp div n}
+//
+// TODO(junbeomlee): Add Frobenius(x, e) for φᵉ(x) = x^{pᵉ} when implementing
+// quartic extension.
+
+// Base case: prime field Frobenius is identity (x^p = x in Fp).
+template <typename T, std::enable_if_t<T::ExtensionDegree() == 1>* = nullptr>
+[[nodiscard]] T Frobenius(const T& x) {
+  return x;
+}
+
+// Extension field case.
+template <typename T, std::enable_if_t<(T::ExtensionDegree() > 1)>* = nullptr>
+[[nodiscard]] T Frobenius(const T& x) {
+  using BaseField = typename ExtensionFieldOperationTraits<T>::BaseField;
+  using BasePrimeField = typename T::Config::BasePrimeField;
+  constexpr size_t kDegree = ExtensionFieldOperationTraits<T>::kDegree;
+
+  const auto& p = BasePrimeField::Config::kModulus;
+  BaseField xi = T::Config::kNonResidue;
+
+  std::array<BaseField, kDegree> src = x.ToBaseField();
+  std::array<BaseField, kDegree> dst{};
+
+  // Compute (k * p) mod n and (k * p) div n without overflow.
+  constexpr auto p_div_n = p / kDegree;
+  constexpr auto p_mod_n = p % kDegree;
+
+  for (size_t k = 0; k < kDegree; ++k) {
+    size_t kr = k * p_mod_n;
+    size_t new_pos = kr % kDegree;
+    auto exp = k * p_div_n + kr / kDegree;
+
+    BaseField coeff = (k == 0) ? BaseField::One() : xi.Pow(exp);
+    // x_k · u^k → φ(x_k) · ξ^{(kp) div n} · u^{(kp) mod n}
+    dst[new_pos] = dst[new_pos] + Frobenius(src[k]) * coeff;
+  }
```

Comment:
Could you change the computation like https://fractalyze.gitbook.io/intro/primitives/abstract-algebra/extension-field/inversion#id-2.2.-optimized-computation-when?